### ISSUE TITLE
Allow TaggedLocator autowiring via enums

### ISSUE THREAD

**Issue by zmitic** [symfony/symfony#44556]
### Description

```php
#[TaggedLocator(tag: 'my_tag', defaultIndexMethod: 'myStaticMethod')]
```

works with strings and integers, but not with ``BackedEnum``. A simple change to 

```php
trait PriorityTaggedServiceTrait
{
    public static function getDefault()
    {

        // ....
        if ($default instanceof BackedEnum) {    //  <---
            $default = $default->value;
        }

        if (\is_int($default)) {
            $default = (string) $default;
        }
        // ....
        
    }
}
```

could improve static analysis and autocomplete.

### Example

Tag:

```php
#[AutoconfigureTag(name: 'my_tag')]
interface PartnerInterface
{
    public static function getPartnerId(): MyEnum;   // <- enum instead of string
}
```

Service that gets them works as before:

```php
class PartnerProcessor
{
    public function __construct(
        #[TaggedLocator(tag: 'my_tag', defaultIndexMethod: 'getPartnerId')]
        private ServiceLocator $partners,
    )
    {
    }

    // autocomplete and static analysis is improved here
    public function doSomethingWithPartner(MyEnum $enum): void
    {
        $partner = $this->partners->get($enum->value);
    }
}
```

**Comment by nicolas-grekas** [comment]
IIUC, while nothing would enforce it, this would allow creating a convention to limited list of possible service ids, isn't it?
Would you like sending a PR to give this a try?

**Comment by zmitic** [comment]
Yes, gladly. Expect by Wednesday, just to get around writing tests and keep BC with <8.1.

**Comment by zmitic** [comment]
@nicolas-grekas I made PR but given this is the first one, I probably made some mistakes in the process. I would really appreciate if you could ELI5 them so I don't repeat them in the future with more complex features.

**Comment by stof** [comment]
Well, this proposal is using enums as a way to store backing strings that are used to actually index the service locator. This seems like an abuse of enums. That's not how enums are expected to be used (enums are providing type safety if you use the enum, not its backing value). Here, you are still using the string when reading from the service locator

**Comment by zmitic** [comment]
@stof 
That's a good point, I didn't think of that. 

But `Symfony\Component\DependencyInjection\ServiceLocator::get(string $id)` wouldn't allow me to pass enum case. I could remove that typehint and cast `$id` to string (with check for >=8.1); do you think it is a good idea to do?

**Comment by stof** [comment]
We can't. That `get` method is defined by psr/container, and we don't want to deviate from that.
